plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/config.py

```diff
@@ -16,35 +17,85 @@ class QueueType(Enum):
 class SlurmQueueConf:
     # Params are used to configure sbatch, for more info check:
     # https://github.com/facebookincubator/submitit/blob/master/submitit/slurm/slurm.py
-    nodes: int = 1
+
+    # maximum time for the job in minutes
+    time: int = 60
+    # number of cpus to use for each task
+    cpus_per_task: int = 10
+    # number of gpus to use on each node
     gpus_per_node: int = 1
+    # number of tasks to spawn on each node
     ntasks_per_node: int = 1
+    # number of nodes to use for the job
+    nodes: int = 1
+    # memory to reserve for the job on each node
     mem: str = "${hydra.launcher.mem_limit}GB"
```

Comment:
Can you put a comment on mem_limit definition in the launcher that it's in GB?

Reply:
1. I have no idea where this mem_limit is supposed to come from, grepping mem_limit yields nothing in hydra.
2. Is there a way to get this value as an integer and not a string? That would be useful to set it in auto executor as well, since in that case it is not a string.

Reply:
1. https://github.com/facebookresearch/hydra/blob/979024ea459a72334bb733a033514c45d44e39f2/plugins/hydra_submitit_launcher/hydra_plugins/hydra_submitit_launcher/config.py#L70

2. Historically the submitit mem param was in the form xxxGB, what is the recommended form and type for mem?

Reply:
1. Oh I did not realize that would be here thanks
2. the slurm executor accepts the slurm `mem` parameter, so a string which ca be xxxGB or xxxMB etc. We wanted to simplify and further type it in the auto executor, hence it is an integer, and always in GB. In my opinion the recommended form is using auto executor :D

Reply:
This config is for the slurm queue so it makes sense that it passes the expected string (xxxGB).

Are you suggesting that we remove support for the local and slurm queues in favor of auto?
In my opinion you will discover that having a single queue to configure many others will become unsustainable if you add additional backends.

Hydra can better handle this kind of variability through config group options.

Reply:
> Are you suggesting that we remove support for the local and slurm queues in favor of auto?

That's what I have stating from the start :D 

> you will discover that having a single queue to configure many others will become unsustainable if you add additional backends.

I'd say it the other way round, if we had many backend and one would require writting nearly similar configurations for each one, it can be error prone. the aim with auto executor is to share the parameters which are more widely used (currently: gpus_per_nodes, nodes, tasks_per_nodes, name, cpus per tasks, time out). Each executor may have a different way of handling this independently (for now each specific executor has been sticking close to the cluster parameter name), but the auto executor allows unification.
At some point we could support 2 clusters, say "slurm" and "cluster2", most of the parameters we would use would be the ones mentioned above. Only 2 things were different if we wanted to run a job: slurm_partition/cluster2_hostgroup (same purpose, each with cluster specific options), and the `folder` for the xp (because the storage structure was different).

So the same code would run on both clusters without requiring duplicated options.

If ever we need to override one shared option, we could with something like `slurm_gpus_per_node` for instance. Not sure it currently works (I'll crosscheck) but that is one of the things @gwenzek changes have recently enabled.
I know that it looks similar to having different configurations as hydra enables, the thing is, in my opinion with the executors, 99% of times you would want the *equivalent* options on the other cluster, which auto executor can convert automatically

Reply:
> That's what I have stating from the start :D

Not really, you have been saying people should use auto. not to remove slurm and local.

> I'd say it the other way round, if we had many backend and one would require writting nearly similar configurations for each one, it can be error prone

This can be addressed through interpolation. but we need to decide on the exact recommended structure. Once we have a standard individual launchers can translate from the standard to the what is expected by the system it's using. This also makes it easy for people to switch between different launchers (fairtask).

Here is a proposal:
1. Hydra adds job specific resources it the core, containing the common things: cpus, memory, gpus.
2. Submitit creates a SLRUM config node that can potentially be shared with other SLURM launchers.

I am going to redefine the names under the assumption that:
* A node is a host running multiple tasks
* A task is a specific run with resources requirements

Job specific resources in Hydra core config (a job here represent a single invocation of the user function, this is existing Hydra terminology):

These can be shared by all non trivial launchers (submitit, fairtask, ray, joblib):

```yaml
hydra:
  job:
    # Job requested resources.
    # Not all launchers supports these.
    resourecs:
      cpu_cores: 1
      max_mem_gb: 10
      num_gpus: 0
```

Common SLURM config.
Some things are inheriting the values from hydra.job.resources.
If we end up using this for other SLRUM launchers (fairtask), we can promote it out of the submitit plugin.

slurm_common.yaml
```yaml
hydra:
  slurm:
    # number of nodes to allocate for the job
    num_nodes: 1
    # using job to align with Hydra terminology, we can use task too.
    # Number of user function runs on the each node
    jobs_per_node : 1
    gpus_per_job: ${hydra.job.resoures.num_gpus}
    cpu_cores_per_job: ${hydra.job.resoures.cpu_cores}
    job_timeout_minutes: 1
```
hydra/launcher/submitit_slurm.yaml:
```yaml
hydra:
  launcher:
    cls: SubmititSlurmLauncher
    params:
      num_nodes: ${hydra.slurm.num_nodes}
      jobs_per_node : ${hydra.slurm.jobs_per_node}
      gpus_per_job: ${hydra.slurm.gpus_per_job}
      cpu_cores_per_job: ${hydra.slurm.cpu_cores_per_job}
      job_timeout_minutes: ${hydra.slurm.job_timeout_minutes}
```

hydra/launcher/submitit_local.yaml:
```yaml
hydra:
  launcher:
    cls: SubmititLocalLauncher
    params:
      # I don't think the local launcher actually supports these things, but let's pretend.
      # Since local is not SLURM, this can use directly from hydra.job.resources:
      num_gpus: ${hydra.job.resoures.num_gpus}
      cpu_cores: ${hydra.job.resoures.cpu_cores}
```

Thoughts?

Reply:
I was considering something somehow equivalent:
1. Only use AutoExecutor, since it encapsulate everything
2. Move all the shared config (gpus_per_node, mem_gb etc) to one base config (this basically looks like your hydra-scope config
3. Having separate config classes for each cluster (that would basically only be slurm right now) with cluster specific options and *possibly* some overrides of the basic config. In practice I would definitely not like keeping all options there, so users should be able to add parameters that are not present in it.

Considering your proposition, the names are incorrect in my opinion. A job has the similar definition in both hydra and slurm, you're missing the distinction of multi-tasks and multi-node jobs. Currently that is:

```yaml
hydra:
  job:
    # Job requested resources.
    # Not all launchers supports these.
    resources:
      nodes: 1
      tasks_per_node: 1
      cpus_per_task: 1
      mem_gb: 10
      gpus_per_node: 0
      timeout_min: 10
```
`nodes` and `tasks_per_node` are necessary to run multi-node and multi-task jobs so the distinction is important and the need to define cpus and gpus in consequence. The only element I would rename (but for some minor name clarifications) is `gpus_per_node` -> `gpus_per_task`. We have not done it yet for historical reasons and to avoid introducing too many changes with the open source release, but it's planned.

Reply:
> Having separate config classes for each cluster

This becomes trivial with Structured Configs.

> Considering your proposition, the names are incorrect in my opinion. A job has the similar definition in both hydra and slurm, you're missing the distinction of multi-tasks and multi-node jobs

Hydra does not support multi-task (what is that) or multi-node jobs.
Those are properties of the Submitit (and potentially other SLURM based launchers).

Reply:
> In my proposal there is also no Hydra support for cpu, mem and gpus

I don't get what you mean, that looks exactly like it does.

How do you make interpolation work with dataclasses when you need for instance the multiplication of 2 parameters as an integer as well?

Reply:
> I don't get what you mean, that looks exactly like it does.

Where does it look like that?
If the answer is in the above snippet, this is not something in Hydra now.
and even if it gets in, it will only mean that a job can declare it's resource requirements. different launchers would still need to respect that.

I think details like supporting multiple jobs on the same node (multitask) or not on different nodes (multi-node?) are details that are more specific to SLRUM which are not as generic.

> How do you make interpolation work with dataclasses when you need for instance the multiplication of 2 parameters as an integer as well?

Currently OmegaConf does not support arithmetic operations.
This is something I will consider in a future version ([task](https://github.com/omry/omegaconf/issues/91)).

In this case, this kind of logic can be handled by the Launcher itself before it's calling submitit.

Reply:
> In this case, this kind of logic can be handled by the Launcher itself before it's calling submitit.

I would not be the one doing that, I hate duplicated work. In my opinion after getting more understanding of the plugin, this is done and should stay in the submitit package, and this plugin should only use the AutoExecutor. I'll finish this basic PR without changing the behavior for now anyway, let's chat about this offline

Reply:
I didn't plan for these changes in this diff. Changes to Hydra are frozen except bug fixes until I start working on 1.1.

(We can of course freely change the plugin).
I agree that we should continue discussing this offline.